File: gemsfx/src/main/java/com/dlsc/gemsfx/YearMonthView.java
package com.dlsc.gemsfx;

import com.dlsc.gemsfx.skins.YearMonthViewSkin;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.css.CssMetaData;
import javafx.css.Styleable;
import javafx.css.StyleableBooleanProperty;
import javafx.css.StyleableProperty;
import javafx.css.converter.BooleanConverter;
import javafx.scene.control.Control;
import javafx.scene.control.Skin;
import javafx.util.StringConverter;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import java.time.Month;
import java.time.YearMonth;
import java.time.format.TextStyle;
import java.util.Locale;
import java.util.Objects;

/**
 * A view for displaying and selecting year-month values.
 *
 * <p><b>CSS Styleable Properties:</b>
 * <table class="striped">
 *   <caption>CSS Properties</caption>
 *   <thead><tr><th>Property</th><th>Type</th><th>Description</th></tr></thead>
 *   <tbody>
 *     <tr><td>{@code -fx-show-year}</td><td>{@code Boolean}</td><td>Whether to show the year at the top.</td></tr>
 *   </tbody>
 * </table>
 */
public class YearMonthView extends Control {

    /**
     * Constructs a new year-month view.
     */
    public YearMonthView() {
        getStyleClass().add("year-month-view");
        setFocusTraversable(false);

        // the UI / the skin should prevent this from happening, but just in case, we make sure that the value is
        // always within the allowed range
        valueProperty().addListener((obs, oldValue, newValue) -> {
            if (newValue != null) {
                YearMonth earliestMonth = getEarliestMonth();
                YearMonth latestMonth = getLatestMonth();
                if (earliestMonth != null && newValue.isBefore(earliestMonth)) {
                    setValue(earliestMonth);
                } else if (latestMonth != null && newValue.isAfter(latestMonth)) {
                    setValue(latestMonth);
                }
            }
        });
    }

    /**
     * {@inheritDoc}
     *
     * @return the default skin
     */
    @Override
    protected Skin<?> createDefaultSkin() {
        return new YearMonthViewSkin(this);
    }

    /**
     * {@inheritDoc}
     *
     * @return the user agent stylesheet
     */
    @Override
    public String getUserAgentStylesheet() {
        return Objects.requireNonNull(YearMonthView.class.getResource("year-month-view.css")).toExternalForm();
    }

    private final StyleableBooleanProperty showYear = new StyleableBooleanProperty(true) {
        /**
         * {@inheritDoc}
         *
         * @return the owning bean
         */
        @Override
        public Object getBean() { return YearMonthView.this; }
        /**
         * {@inheritDoc}
         *
         * @return the property name
         */
        @Override
        public String getName() { return "showYear"; }
        /**
         * {@inheritDoc}
         *
         * @return the CSS metadata for this property
         */
        @Override
        public CssMetaData<? extends Styleable, Boolean> getCssMetaData() {
            return StyleableProperties.SHOW_YEAR;
        }
    };

    public final boolean isShowYear() {
        return showYear.get();
    }

    /**
     * Determines if the view willl display the current year at the top or not. When this
     * control is used standalone, then showing the year is usually necessary.
     * If the control is part of a more complex control like the {@link CalendarView} then
     * showing the year might not be needed as it becomes obvious from the overall context.
     * <p>
     * Can be set via CSS using the {@code -fx-show-year} property.
     * Valid values are: {@code true}, {@code false}.
     * The default value is {@code true}.
     * </p>
     *
     * @return true if the year will be shown
     */
    public final BooleanProperty showYearProperty() {
        return showYear;
    }

    public final void setShowYear(boolean showYear) {
        this.showYear.set(showYear);
    }

    private static class StyleableProperties {

        private static final CssMetaData<YearMonthView, Boolean> SHOW_YEAR =
            new CssMetaData<>("-fx-show-year", BooleanConverter.getInstance(), true) {
                /**
                 * {@inheritDoc}
                 *
                 * @return true if the property can be styled
                 *
                 * @param c the control to inspect
                 */
                @Override
                public boolean isSettable(YearMonthView c) {
                    return !c.showYear.isBound();
                }
                /**
                 * {@inheritDoc}
                 *
                 * @return the styleable property
                 *
                 * @param c the control to inspect
                 */
                @Override
                public StyleableProperty<Boolean> getStyleableProperty(YearMonthView c) {
                    return (StyleableProperty<Boolean>) c.showYear;
                }
            };

        private static final List<CssMetaData<? extends Styleable, ?>> STYLEABLES;
        static {
            List<CssMetaData<? extends Styleable, ?>> styleables = new ArrayList<>(Control.getClassCssMetaData());
            styleables.add(SHOW_YEAR);
            STYLEABLES = Collections.unmodifiableList(styleables);
        }
    }

    /**
     * Returns the CSS metadata supported by this control.
     *
     * @return the CSS metadata supported by this control
     */
    public static List<CssMetaData<? extends Styleable, ?>> getClassCssMetaData() {
        return StyleableProperties.STYLEABLES;
    }

    /**
     * {@inheritDoc}
     *
     * @return the supported CSS metadata
     */
    @Override
    public List<CssMetaData<? extends Styleable, ?>> getControlCssMetaData() {
        return getClassCssMetaData();
    }

    private final ObjectProperty<YearMonth> value = new SimpleObjectProperty<>(this, "value", YearMonth.now());

    public final YearMonth getValue() {
        return value.get();
    }

    /**
     * Stores the selected year and month.
     *
     * @return the selected year-month property
     */
    public final ObjectProperty<YearMonth> valueProperty() {
        return value;
    }

    public final void setValue(YearMonth value) {
        this.value.set(value);
    }

    private final ObjectProperty<StringConverter<Month>> converter = new SimpleObjectProperty<>(this, "converter", new StringConverter<>() {
        /**
         * {@inheritDoc}
         *
         * @return the string representation of the value
         *
         * @param month the value to convert
         */
        @Override
        public String toString(Month month) {
            if (month != null) {
                return month.getDisplayName(TextStyle.FULL, Locale.getDefault());
            }
            return null;
        }

        /**
         * {@inheritDoc}
         *
         * @return the parsed value
         *
         * @param s the string to parse
         */
        @Override
        public Month fromString(String s) {
            return null;
        }
    });

    public final StringConverter<Month> getConverter() {
        return converter.get();
    }

    /**
     * Stores the converter used to format the month names shown by the view.
     *
     * @return the month converter property
     */
    public final ObjectProperty<StringConverter<Month>> converterProperty() {
        return converter;
    }

    public final void setConverter(StringConverter<Month> converter) {
        this.converter.set(converter);
    }

    private final ObjectProperty<YearMonth> earliestMonth = new SimpleObjectProperty<>(this, "earliestMonth");

    public final YearMonth getEarliestMonth() {
        return earliestMonth.get();
    }

    /**
     * Determines the earliest month that the user can select in the view.
     *
     * @return the earliest selectable month
     */
    public final ObjectProperty<YearMonth> earliestMonthProperty() {
        return earliestMonth;
    }

    public final void setEarliestMonth(YearMonth earliestMonth) {
        this.earliestMonth.set(earliestMonth);
    }

    private final ObjectProperty<YearMonth> latestMonth = new SimpleObjectProperty<>(this, "latestMonth");

    public final YearMonth getLatestMonth() {
        return latestMonth.get();
    }

    /**
     * Determines the latest month that the user can select in the view.
     *
     * @return the latest selectable month
     */
    public final ObjectProperty<YearMonth> latestMonthProperty() {
        return latestMonth;
    }

    public final void setLatestMonth(YearMonth latestMonth) {
        this.latestMonth.set(latestMonth);
    }
}

File: docs/manuals/src/manuals/year_month_view.py
"""Content of the YearMonthView developer manual."""

from manualkit import (
    Bullets, Callout, Chapter, Code, Figure, Manual, Numbered, Para,
    Property, PropertyTable, Section, Table,
)

G = "year-month-view"

MANUAL = Manual(
    control="YearMonthView",
    package="com.dlsc.gemsfx",
    subtitle="A two-column month grid for selecting YearMonth values",
    abstract=("YearMonthView displays the twelve months of a year as selectable cells, optionally bounded by earliest and latest months. It is useful as a standalone month selector and as popup content inside richer date controls."),
    cover_svg=f"{G}/cover.svg",
    cover_caption="Generated cartoon overview of YearMonthView.",
    chapters=[
        Chapter("Introduction", [
            Para("<b>YearMonthView</b> is part of the GemsFX module <font face='Courier'>com.dlsc.gemsfx</font>. This manual documents only behavior verified in the control, skin, CSS, resource bundles and demo app."),
            Section("Key features"),
            Bullets(['Selects a java.time.YearMonth value, initialized to YearMonth.now().', 'Header with previous / next year buttons can be shown or hidden.', 'Month names are formatted by a StringConverter<Month>.', 'Optional earliestMonth and latestMonth disable months outside the valid range.', 'Selected and current months receive pseudo-class state in the skin.']),
            Section("Maven dependency"),
            Code("""<dependency>
    <groupId>com.dlsc.gemsfx</groupId>
    <artifactId>gemsfx</artifactId>
    <version>4.4.1</version>
</dependency>""", caption="Use package <font face='Courier'>com.dlsc.gemsfx</font>."),
        ]),
        Chapter("Getting started", [
            Para("Create the control, set the properties that define its valid values, and listen to the value or selection property that the control owns."),
            Code("""YearMonthView view = new YearMonthView();
view.setEarliestMonth(YearMonth.now().minusYears(1));
view.setLatestMonth(YearMonth.now());
view.setMaxSize(Region.USE_PREF_SIZE, Region.USE_PREF_SIZE);

view.valueProperty().addListener((obs, oldValue, newValue) -> {
    System.out.println("selected month = " + newValue);
});""", caption="A compact setup for <font face='Courier'>YearMonthView</font>."),
            Figure(f"{G}/cover.svg", "A first look at the control in a simple application window."),
        ]),
        Chapter("Anatomy", [
            Para("The skin builds the visible nodes below the public control. The table lists the style classes and nodes that are useful when reading the source or writing CSS."),
            Figure(f"{G}/anatomy.svg", "The main parts of the control."),
            Table(["Part", "Style / node", "Description"], [
                        ["Root", "year-month-view", "Control root and stylesheet hook."],
                        ["Container", "container", "VBox containing the optional header and month grid."],
                        ["Header", "header", "Left arrow, year label and right arrow; hidden when showYear is false."],
                        ["Grid", "grid-pane", "Six rows and two month columns separated by a divider."],
                        ["Month cell", "month-box / month-label", "Selectable month label created from converter.toString(month)."],
                        ["Indicator", "indicator", "Visible for the selected month via the selected pseudo class."]
            ], widths=[18, 28, 54]),
        ]),
        Chapter("Control API", [
            Section("Selection"),
            PropertyTable([
                        Property("value", "ObjectProperty&lt;YearMonth&gt;", "YearMonth.now()", "Selected year and month. Programmatic changes are constrained to earliestMonth/latestMonth when those bounds exist."),
                        Property("earliestMonth", "ObjectProperty&lt;YearMonth&gt;", "null", "Earliest selectable month; earlier month boxes are disabled."),
                        Property("latestMonth", "ObjectProperty&lt;YearMonth&gt;", "null", "Latest selectable month; later month boxes are disabled.")
            ]),
            Section("Display"),
            PropertyTable([
                        Property("showYear", "BooleanProperty", "true", "Shows the year header. Styleable via -fx-show-year."),
                        Property("converter", "ObjectProperty&lt;StringConverter&lt;Month&gt;&gt;", "full month name for default Locale", "Formats the month names shown in each month box. fromString returns null in the default converter.")
            ]),
        ]),
        Chapter("Behaviour", [
            Section("Bounded month selection"),
            Para("Changing value, earliestMonth or latestMonth keeps the selected value inside the allowed interval. The arrows are disabled at the boundary year and individual month boxes are disabled if their YearMonth is outside the range."),
            Figure(f"{G}/behaviour.svg", "Interaction and value-flow behaviour."),
            Section("Month formatting"),
            Para("The skin creates one label per Month by calling the converter. Use this for abbreviations, uppercase month names or domain-specific labels."),
        ]),
        Chapter("Styling", [
            Para("The stylesheet is loaded by the control or its base class. The rows below list style hooks verified in the CSS and skin source."),
            Figure(f"{G}/styling.svg", "Style classes and pseudo-class states."),
            Section("Style classes"),
            Table(["Style class", "Where used"], [
                        ["year-month-view", "Root control style class."],
                        ["container", "VBox wrapping header and grid."],
                        ["header", "Year navigation row."],
                        ["arrow-button left-button right-button", "Navigation buttons."],
                        ["arrow left-arrow right-arrow", "Arrow glyph regions."],
                        ["grid-pane", "GridPane containing month cells and divider."],
                        ["month-box", "VBox for one month."],
                        ["month-label", "Label created by converter."],
                        ["indicator", "Selection indicator below each month."],
                        ["divider", "Vertical divider between odd and even month columns."],
                        ["popover", "Optional root style class with popup-specific border/background tweaks."]
            ], widths=[35,65]),
            Section("Pseudo classes"),
            Table(["Pseudo class / marker", "Meaning"], [
                        ["selected", "Applied to the selected month-box."],
                        ["current", "Applied to the current month-box when the displayed year is the current year."]
            ], widths=[30,70]),
            Section("Styleable CSS properties"),
            Table(["Property", "Type", "Default", "Description"], [
                        ["-fx-show-year", "Boolean", "true", "Shows or hides the header."]
            ], widths=[28,18,20,34]),
            Code(""".year-month-view {
    -fx-show-year: false;
}
.year-month-view > .container > .grid-pane .month-box:selected > .indicator {
    -fx-background-color: -fx-accent;
}""", caption="Example CSS."),
        ]),
        Chapter("Localization", [
            Para("This control uses <font face='Courier'>ResourceBundleManager</font> for the following keys."),
            Table(["Key", "English default"], [
                        ["accessible.role-description", "month selector"]
            ], widths=[55,45]),
        ]),
        Chapter("Accessibility", [
            Para("The constructor sets AccessibleRole.DATE_PICKER through AccessibilityUtil.setRole and uses the localized role description \"month selector\"."),
        ]),
        Chapter("Recipes", [
            Figure(f"{G}/recipes.svg", "Common configuration recipes."),
            Section("Limit to the last twelve months"),
            Code("""YearMonthView view = new YearMonthView();
view.setEarliestMonth(YearMonth.now().minusMonths(11));
view.setLatestMonth(YearMonth.now());"""),
            Section("Use abbreviated month names"),
            Code("""view.setConverter(new StringConverter<>() {
    @Override public String toString(Month month) {
        return month.getDisplayName(TextStyle.SHORT, Locale.getDefault());
    }
    @Override public Month fromString(String text) { return null; }
});"""),
            Section("Hide the year header"),
            Code("""view.setShowYear(false);"""),
            Section("Bind another control to the selection"),
            Code("""Label label = new Label();
label.textProperty().bind(Bindings.convert(view.valueProperty()));"""),
            Section("Checklist"),
            Numbered(['Set value before applying tight bounds if you want a specific initial month.', 'Use a converter for all visible month text.', 'Do not select disabled months in custom event handlers.', 'Keep the control at USE_PREF_SIZE when using it as compact popup content.']),
        ]),
        Chapter("See also", [
            Para("Demo app: <font face='Courier'>YearMonthViewApp</font>. Run it with:"),
            Code("mvn javafx:run -f gemsfx-demo/pom.xml -Dmain.class=com.dlsc.gemsfx.demo.YearMonthViewApp"),
            Bullets([
                "Related GemsFX controls: YearMonthPicker, YearPicker, YearView, CalendarView.",
                "API documentation: https://dlsc-software-consulting-gmbh.github.io/GemsFX/api/",
            ]),
        ]),
    ],
)
